### PR TITLE
Add wasm-bindgen to the rustc build

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is targeted at enabling by default the `#[wasm_bindgen]` attribute
for Rust code. This involved a few primary pieces to be available on the
builder:

* First the runtime support of the `wasm_bindgen` crate and the
  `#[wasm_bindgen]` macro needed to be available when compiling code.
* Next the `wasm-bindgen` CLI tool needed to be available and automatically run
  whenever a wasm file is generated with an appropriate section name.
* Finally the generated JS as well as the generated wasm needed to be sent back
  to the client

The `/rustc` has been updated with all these steps and transitioned to `async`
and `await` at the same time to make it a bit clearer to see what's going on.
The slug generation has also been updated to primarily go through Docker to
ensure a nicely containerized build. I've deployed this current slug to
https://young-sierra-27135.herokuapp.com/ for inspection as well.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'm still working on integrating this into WebAssemblyStudio itself so this is still WIP, and I think the options for running wasm-bindgen may be tweaked a bit before this is done.

**Comment by alexcrichton** [comment]
No longer WIP!

A demo of this is available at https://alexcrichton.github.io/WebAssemblyStudio/ with this repo published to the URL mentioned in OP

File: app/src/rustc.js [hunk 1]
```diff
@@ -1,44 +1,80 @@
...
-            if (!err) fs.unlink(wasmFile, () => {});
-            fs.unlink(rustFile, () => {});
-            callback(null, { success, output: wasm, console, });
-          });
-        });
+  try {
+    let args = [rustcCmd, rustFile];
+    args.push('--target=wasm32-unknown-unknown');
+    args.push('--crate-type=cdylib');
+    args.push(options);
```
Comment (yurydelendik) [inline app/src/rustc.js]: Can you remove 'options' from here and only use explicit/sanitized values? e.g. `if (options.lto) args.push('-Clto');`

**Review by yurydelendik — APPROVED** [review]

**Comment by alexcrichton** [comment]
Ok updated!

I'll update the templates to enable optimizations and LTO by default in WebAssemblyStudio?

**Comment by yurydelendik** [comment]
Thank you for the patch.

**Pull request merged**